src/layer_lifecycle.rs

```diff
@@ -0,0 +1,367 @@
+
+    /// Tries to recover from invalid layer metadata
+    ///
+    /// When the metadata of the existing layer cannot be deserialized into `LM`, this function will
+    /// be called by [`execute_layer_lifecycle`] with the actual layer metadata as TOML. This allows
+    /// implementors to recover from this case by, for example, implementing migrations from older
+    /// versions of the metadata to the current one.
+    ///
+    /// The default implementation returns [`MetadataRecoveryStrategy::DeleteLayer`] to signal that
+    /// the existing layer should be deleted in its entirety.
+    fn recover_from_invalid_metadata(
+        &self,
+        #[allow(unused_variables)] layer_metadata: &toml::value::Table,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> Result<MetadataRecoveryStrategy<LM>, E> {
+        // Default implementation is to delete the layer if the metadata is invalid
+        Ok(MetadataRecoveryStrategy::DeleteLayer)
+    }
+
+    /// Based on the current state of the layer, determines how the layer will be processed
+    ///
+    /// This will be called by [`execute_layer_lifecycle`] when the layer for this lifecycle already
+    /// exists. Based on the layer contents and its metadata, implementors can decide how the layer
+    /// will be processed.
+    ///
+    /// A common case is the comparison of the layer metadata to some expected value to determine
+    /// if the contents are still valid for the current build. If the contents are no longer valid,
+    /// an implementation might return [`ValidateResult::RecreateLayer`] which will result in the
+    /// layer being deleted and then recreated via [`LayerLifecycle::create`]. If the layer contents
+    /// are still valid, [`ValidateResult::KeepLayer`] could be returned to keep the layer as-is
+    /// without doing any modifications to it.
+    ///
+    /// The common case above is just one conceivable implementation of this function, take a look
+    /// at [`ValidateResult`] for additional validation results and their meanings.
+    fn validate(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
+        #[allow(unused_variables)] layer_content_metadata: &LayerContentMetadata<LM>,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> ValidateResult {
+        // Default implementation is to always recreate the layer
+        ValidateResult::RecreateLayer
+    }
+
+    /// Updates an existing layer
+    fn update(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
+        #[allow(unused_variables)] layer_content_metadata: LayerContentMetadata<LM>,
+        #[allow(unused_variables)] build_context: &BuildContext<P, BM>,
+    ) -> Result<LayerContentMetadata<LM>, E> {
+        // Default implementation is a no-op
+        Ok(layer_content_metadata)
+    }
+
+    fn layer_lifecycle_data(
+        &self,
+        #[allow(unused_variables)] layer_path: &Path,
```

Comment:
```suggestion
        _layer_path: &Path,
```

Reply:
I use the attributes here to ensure that both the docs and generated code from IDEs (or other tooling) does not create parameters prefixed with `_`.